### PR TITLE
BUG: CategoricalIndex allowed reindexing duplicate sources

### LINKED ISSUE #25459: CategoricalIndex.reindex raises on duplicate indexer
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

# works fine
x = pd.Series([0])
x.reindex([0,0])

# ValueError: cannot reindex with a non-unique indexer
x.index = x.index.astype('category')
x.reindex([0,0])
```

#### Problem description

When a series or frame has a categorical index, it cannot be reindexed using an indexer that has duplicate values. This is inconsistent with other indexes, and makes categorical indexes less useful.

#### Expected Output

No error.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Linux
OS-release: 4.19.15-200.fc28.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.24.1
pytest: 4.3.0
pip: 19.0.3
setuptools: 40.8.0
Cython: 0.29.5
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 4.2.1
sphinx: None
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: 2.6.2
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.3.1
bs4: 4.7.1
html5lib: None
sqlalchemy: 1.2.18
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by batterseapower**
For consistency with normal indexes, `CategoricalIndex.reindex` should allow targets to be duplicated but not the sources. However, currently in allows the source to be duplicated but not the targets, which is exactly the wrong behaviour.

Most of the work here is fixing the tests, which in many cases explicitly check for the incorrect behaviour.

Fixes #25459 

(I can't run the full testsuite on my machine but the relevant categorical tests do seem to pass. Hoping that the GitHub CI infrastructure will pick up any other failures.)

### CODE REVIEW

File: pandas/tests/indexes/test_category.py [hunk 1]
```diff
@@ -574,41 +574,56 @@ def test_reindexing(self):
             tm.assert_numpy_array_equal(expected, actual)
 
     def test_reindex_dtype(self):
-        c = CategoricalIndex(["a", "b", "c", "a"])
+        c = CategoricalIndex(["a", "b", "c"])
```
Comment (TomAugspurger) [inline pandas/tests/indexes/test_category.py]: Why these changes? Was the previous behavior incorrect?

I'd prefer to keep the diff in the tests as small as possible so that the changes here are easier to understand.

Comment (batterseapower) [reply pandas/tests/indexes/test_category.py]: You are right, I might actually have changed this unnecessarily. The changes to the `DataFrame.reindex` tests *are* necessary, but these changes might not be.

**Comment by batterseapower** [comment]
Had to fix a couple of incidental Pandas bugs that were surfaced by the main fix.

1) `pd.Index([1, 1, 0, 2, 2]).union(pd.Index([0, 2, -1]))` would fail with an error about duplicates in the index. We now return `pd.Index([1, 1, 0, 2, 2, -1])` which is more consistent with the behaviour of `Index.intersection`

2) `pd.Index(['A', 'B']).get_indexer_non_unique(pd.Index([0]))` would fail, complaining with a `TypeError` about `'<' not supported between instances of 'str' and 'int'`. This is caused by over-aggressive use of `searchsorted`, solved by falling back to linear search

File: doc/source/whatsnew/v1.0.0.rst [hunk 2]
```diff
@@ -143,6 +143,8 @@ Indexing
 ^^^^^^^^
 
 - Bug in assignment using a reverse slicer (:issue:`26939`)
+- :meth:`Index.union` could fail when the LHS contained duplicates (:issue:`28257`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: LHS -> left

File: pandas/_libs/index.pyx [hunk 3]
```diff
@@ -304,12 +304,20 @@ cdef class IndexEngine:
         if stargets and len(stargets) < 5 and self.is_monotonic_increasing:
             # if there are few enough stargets and the index is monotonically
             # increasing, then use binary search for each starget
+            remaining_stargets = set()
```
Comment (jreback) [inline pandas/_libs/index.pyx]: can you add this to the cdef above

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -2489,8 +2489,14 @@ def _union(self, other, sort):
                 value_set = set(lvals)
                 result.extend([x for x in rvals if x not in value_set])
         else:
-            indexer = self.get_indexer(other)
-            indexer, = (indexer == -1).nonzero()
+            # find indexes of things in "other" that are not in "self"
```
Comment (jreback) [inline pandas/core/indexes/base.py]: rather than try/except you can just check ``other.is_unique`` first

File: pandas/tests/indexing/test_categorical.py [hunk 5]
```diff
@@ -46,6 +46,18 @@ def setup_method(self, method):
                 "B": (Series([1, 1, 2, 1, 3, 2]).astype(CDT([3, 2, 1], ordered=False))),
             }
         ).set_index("B")
+        self.df5 = DataFrame(
```
Comment (jreback) [inline pandas/tests/indexing/test_categorical.py]: rather than adding to something that needs to be changed (e.g. we need to paramterize these), create a fixture or just create in-line

File: pandas/tests/indexing/test_categorical.py [hunk 6]
```diff
@@ -564,89 +576,82 @@ def test_reindexing(self):
 
         # reindexing
         # convert to a regular index
```
Comment (jreback) [inline pandas/tests/indexing/test_categorical.py]: make a new test function(s), create the inputs in-line

File: pandas/tests/series/test_operators.py [hunk 7]
```diff
@@ -257,8 +256,8 @@ def test_logical_ops_with_index(self, op):
             pytest.param(
                 ops.ror_,
                 marks=pytest.mark.xfail(
-                    reason="Index.get_indexer with non unique index",
-                    raises=InvalidIndexError,
+                    reason="GH#22092 Index implementation returns Index",
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: huh?

File: pandas/tests/test_base.py [hunk 8]
```diff
@@ -990,6 +990,11 @@ def test_bool_indexing(self, indexer_klass, indexer):
             s = pd.Series(idx)
             tm.assert_series_equal(s[indexer_klass(indexer)], s.iloc[exp_idx])
 
+    def test_get_indexer_non_unique_dtype_mismatch(self):
+        indexes, missing = pd.Index(["A", "B"]).get_indexer_non_unique(pd.Index([0]))
```
Comment (jreback) [inline pandas/tests/test_base.py]: add the issue number

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexing/test_categorical.py (hunk 5 above)
Comment (batterseapower) [reply pandas/tests/indexing/test_categorical.py]: I've moved this to be created inline

File: pandas/tests/indexing/test_categorical.py (hunk 6 above)
Comment (batterseapower) [reply pandas/tests/indexing/test_categorical.py]: The test is quite big, but I didn't make it any bigger. I could split it up at the cost of duplicating a bit of code (mostly just the test data). Is that what you're after?

File: doc/source/user_guide/advanced.rst [hunk 9]
```diff
@@ -783,10 +783,10 @@ values **not** in the categories, similarly to how you can reindex **any** panda
 
 .. ipython:: python
 
-   df2.reindex(['a', 'e'])
-   df2.reindex(['a', 'e']).index
-   df2.reindex(pd.Categorical(['a', 'e'], categories=list('abcde')))
-   df2.reindex(pd.Categorical(['a', 'e'], categories=list('abcde'))).index
+   df2.iloc[[0]].reindex(['a', 'e'])
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst]: this is a very confusing example now.  I would just construct a new frame and use that as what I care about as a reader is what does ``df.reindex(....)`` do

File: doc/source/whatsnew/v1.0.0.rst [hunk 10]
```diff
@@ -89,7 +89,7 @@ Categorical
 ^^^^^^^^^^^
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
--
+- For :class:`CategoricalIndex`, `DataFrame.reindex` would fail when the targets contained duplicates, and wouldn't fail if the source contained duplicates (:issue:`28107`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: use ``:meth:`DataFrame.reindex` ``

Comment (jreback) [reply doc/source/whatsnew/v1.0.0.rst]: :meth:`DataFrame.reindex` with a :class:`CategoricalIndex`, would fail when the targets contained duplicates, and wouldn't fail if the source contained duplicates (:issue:`28107`)

File: doc/source/whatsnew/v1.0.0.rst [hunk 11]
```diff
@@ -143,6 +143,8 @@ Indexing
 ^^^^^^^^
 
 - Bug in assignment using a reverse slicer (:issue:`26939`)
+- :meth:`Index.union` could fail when the left contained duplicates (:issue:`28257`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: this is a general patch on .union?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.0.0.rst (hunk 11 above)
Comment (batterseapower) [reply doc/source/whatsnew/v1.0.0.rst]: Yes.

File: doc/source/whatsnew/v1.0.0.rst (hunk 10 above)
Comment (batterseapower) [reply doc/source/whatsnew/v1.0.0.rst]: Done

File: doc/source/user_guide/advanced.rst (hunk 9 above)
Comment (batterseapower) [reply doc/source/user_guide/advanced.rst]: Agreed, I've changed it.

File: doc/source/whatsnew/v1.0.0.rst (hunk 2 above)
Comment (TomAugspurger) [reply doc/source/whatsnew/v1.0.0.rst]: Can you fix this?

File: doc/source/whatsnew/v1.0.0.rst [hunk 12]
```diff
@@ -106,7 +106,7 @@ Categorical
 ^^^^^^^^^^^
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
--
+- For :class:`CategoricalIndex`, :meth:`DataFrame.reindex` with a :class:`CategoricalIndex`, would fail when the targets contained duplicates, and wouldn't fail if the source contained duplicates (:issue:`28107`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: Does the first "For class:CategoricalIndex:" add anything?

File: pandas/util/testing.py [hunk 13]
```diff
@@ -1736,7 +1736,9 @@ def makeUnicodeIndex(k=10, name=None):
 def makeCategoricalIndex(k=10, n=3, name=None, **kwargs):
     """ make a length k index or n categories """
     x = rands_array(nchars=4, size=n)
-    return CategoricalIndex(np.random.choice(x, k), name=name, **kwargs)
+    return CategoricalIndex(
```
Comment (TomAugspurger) [inline pandas/util/testing.py]: Why this change?

**Review by TomAugspurger — COMMENTED** [review]
Looking good overall.

File: doc/source/whatsnew/v1.0.0.rst (hunk 2 above)
Comment (batterseapower) [reply doc/source/whatsnew/v1.0.0.rst]: Sorry, I did fix this at one point but I lost it in a merge. Should be OK now.

File: pandas/util/testing.py (hunk 13 above)
Comment (batterseapower) [reply pandas/util/testing.py]: Primarily because this method is being used to generate some a `CategoricalIndex` in a parameterised fixture, which is then used as an input to `reindex`. If using `np.random.choice` this test will usually get a duplicate index and hence fail, while `np.arange` won't cause this issue because the test is asking for a short-enough index.

Note that most of the other `Index`-generating methods in this module (in particular `makeIntIndex`, `makeUIntIndex`) avoid creating duplicate indexes. Even `makeStringIndex`, `makeUnicodeIndex`, `makeFloatIndex` avoid creating duplicate indexes with high probability and `makeBoolIndex` seems to be trying to do the same thing if you ask for a short enough index.

File: pandas/core/indexes/base.py [hunk 14]
```diff
@@ -2489,8 +2489,12 @@ def _union(self, other, sort):
...
+            if self.is_unique:
+                indexer = self.get_indexer(other)
+                indexer, = (indexer == -1).nonzero()
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py]: i dont think we use the trailing comma pattern often.  Can you use `[0]` instead?

File: pandas/tests/indexes/test_category.py [hunk 15]
```diff
@@ -598,17 +598,32 @@ def test_reindex_dtype(self):
...
-            c.reindex(
-                CategoricalIndex(["a", "a", "c"], categories=["a", "b", "c", "d"])
+        with pytest.raises(ValueError, match="duplicate axis"):
+            c._can_reindex(
+                CategoricalIndex(["a", "c"], categories=["a", "b", "c", "d"])
             )
 
+    def test_reindex_duplicate_target(self):
+        # See GH25459
+        c = CategoricalIndex(["a", "b", "c"], categories=["a", "b", "c", "d"])
```
Comment (jbrockmendel) [inline pandas/tests/indexes/test_category.py]: i have a pet peeve about 1-letter variable names.  can you make this e.g. `cat`.  same above

File: pandas/tests/indexing/test_categorical.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_categorical.py]: not necessarily part of this PR, but if this test can be split up into individual tests with clearer targets that'd be  helpful

File: pandas/tests/series/test_operators.py [hunk 16]
```diff
@@ -275,41 +274,24 @@ def test_logical_ops_with_index(self, op):
...
-            ),
-            pytest.param(
-                ops.ror_,
-                marks=pytest.mark.xfail(
-                    reason="Index.get_indexer with non unique index",
-                    raises=InvalidIndexError,
-                    strict=True,
-                ),
-            ),
-            ops.rxor,
```
Comment (jbrockmendel) [inline pandas/tests/series/test_operators.py]: why is this change needed?  I guess the ror_ reason would no longer be accurate, but this doesn't change the fact that `op(index, series)` doesn't behave like `op(series, index)`, which is why we have these xfailed

File: pandas/core/indexes/base.py (hunk 14 above)
Comment (batterseapower) [reply pandas/core/indexes/base.py]: I think this is a bad change because not only is your suggested alternative longer, but it also removes a check that the 2nd dimension is a singleton. I'm not sure that your argument for changing it is very compelling. But I don't really want to argue about it, so I'll change it!

File: pandas/tests/indexes/test_category.py (hunk 15 above)
Comment (batterseapower) [reply pandas/tests/indexes/test_category.py]: OK, but please note that this renders my new tests inconsistent with the surrounding ones.

File: doc/source/user_guide/advanced.rst [hunk 17]
```diff
@@ -783,10 +783,13 @@ values **not** in the categories, similarly to how you can reindex **any** panda
...
 .. ipython:: python
 
-   df2.reindex(['a', 'e'])
-   df2.reindex(['a', 'e']).index
-   df2.reindex(pd.Categorical(['a', 'e'], categories=list('abcde')))
-   df2.reindex(pd.Categorical(['a', 'e'], categories=list('abcde'))).index
+   df3 = pd.DataFrame({'A': np.arange(3),
+                       'B': pd.Series(list('abc')).astype('category')})
+   df3 = df3.set_index('B')
+   df3.reindex(['a', 'e'])
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst]: you need to show df3, otherwise its not obvious what you are doing. split this into 2 ipython:: python blocks

File: pandas/tests/indexes/test_category.py [hunk 18]
```diff
@@ -598,17 +598,32 @@ def test_reindex_dtype(self):
...
         tm.assert_numpy_array_equal(indexer, np.array([0, 3, 2], dtype=np.intp))
 
-    def test_reindex_duplicate_target(self):
+    def test_reindex_duplicate_source(self):
         # See GH23963
         c = CategoricalIndex(["a", "b", "c", "a"], categories=["a", "b", "c", "d"])
-        with pytest.raises(ValueError, match="non-unique indexer"):
-            c.reindex(["a", "a", "c"])
+        with pytest.raises(ValueError, match="duplicate axis"):
+            c._can_reindex(["a", "c"])
```
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: don't test internal methods, rather the user facing

File: pandas/tests/series/test_operators.py (hunk 16 above)
Comment (jreback) [reply pandas/tests/series/test_operators.py]: @batterseapower can you respond to this here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_operators.py (hunk 16 above)
Comment (batterseapower) [reply pandas/tests/series/test_operators.py]: OK, the intention of the test was not clear. Perhaps it would have been better to have an explicit `isinstance(result, pd.Series)` check in the test rather than having the test explicitly check that `result` was an Index, which made it look like that was the expected behaviour.

File: pandas/tests/indexes/test_category.py (hunk 18 above)
Comment (batterseapower) [reply pandas/tests/indexes/test_category.py]: I will delete this test entirely then, because `_can_reindex` is called directly by the `BlockManager`, not any of the index machinery itself. `test_reindexing` in `tests/indexing/test_categorical.py` should still cover the relevant code.

File: pandas/tests/series/test_operators.py (hunk 16 above)
Comment (batterseapower) [reply pandas/tests/series/test_operators.py]: I've made the tests clearer and restored the xfail mark.

File: doc/source/user_guide/advanced.rst (hunk 17 above)
Comment (batterseapower) [reply doc/source/user_guide/advanced.rst]: I've tried to clean this up a bit, perhaps it could still be improved though

**Comment by batterseapower** [comment]
Please let me know if this needs more work..

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. cc @jbrockmendel @jschendel if any comments.

**Comment by jbrockmendel** [comment]
LGTM

**Comment by jreback** [comment]
thanks @batterseapower
